File: nativeLib/rary.ohd.pproc/src/GageQCEngine/TEXT/bad_tvalues.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "gageqc_defs.h"
#include "gageqc_types.h"
#include "mpe_log_utils.h"


void read_bad_tvalues(char *fname,int m)

{
extern struct bad_daily_values bad_tvalues[6000];
FILE *fp;
char bfile[100],hb5[10],pc[10],ibuf[100];
char message [ GAGEQC_MESSAGE_LEN ];
float fvalue;
int i,ier,iquart;
char *p;

strcpy(bfile,fname);

fp=fopen(bfile,"r");

if(fp==NULL)
{
	memset(message, '\0', GAGEQC_MESSAGE_LEN);
	sprintf(message,"Could not open file: %s\n",bfile);
	logMessage(message);
        return;
}
else
{
	memset(message, '\0',GAGEQC_MESSAGE_LEN);
	sprintf(message,"Opened file: %s\n",bfile);
	logMessage(message);
}
for(i=0;i<6000;i++) {

       if(bad_tvalues[i].used==1)
                       continue;

        p=fgets(ibuf,80,fp);

        if(p==NULL)
                  break;

        ier=sscanf(ibuf,"%s %s %d %f",hb5,pc,&iquart,&fvalue);

        bad_tvalues[i].used=1;
        strcpy(bad_tvalues[i].hb5,hb5);
        strcpy(bad_tvalues[i].parm,pc);
        bad_tvalues[i].day=m;
        bad_tvalues[i].quart=iquart;
        bad_tvalues[i].fvalue=fvalue;

      }

fclose(fp);

return;

}

void purge_bad_tvalues(int iday)

{

extern struct bad_daily_values bad_tvalues[6000];
int i,k;

for(k=0;k<6;k++) {

for(i=0;i<6000;i++) {

        if(bad_tvalues[i].used==1 && 
           bad_tvalues[i].day==iday &&
           bad_tvalues[i].quart==k)
               bad_tvalues[i].used=0;

      }

}

return;

}


int get_bad_tvalues(int iday,int iquart)
        
{

extern struct bad_daily_values bad_tvalues[6000];
extern struct station * tstation;
extern struct tdata tdata[10];
extern int max_tstations;
int i,j,h;

for(i=0;i<6000;i++) {

      if(bad_tvalues[i].used==0)
                       continue;

      if(bad_tvalues[i].day != iday ||
         bad_tvalues[i].quart != iquart)
                          continue;

        for(j=0;j<max_tstations;j++) {

            if(strcmp(bad_tvalues[i].hb5,tstation[j].hb5)==0 && 
               bad_tvalues[i].parm[4]==tstation[j].parm[4]) {

               if(tdata[iday].stn[j].tlevel2[iquart].qual == 5 &&
                  bad_tvalues[i].fvalue != tdata[iday].stn[j].tlevel1[iquart].data &&
                  tdata[iday].stn[j].tlevel1[iquart].data >= 0) {

                   logMessage("for station %s %f %d\n",tstation[j].hb5,bad_tvalues[i].fvalue,
                     tdata[iday].stn[j].tlevel1[iquart].data);              

                  /* eliminate all bad values for current month */                                  
                   for(h=0;h<6000;h++) {
 
                      if(bad_tvalues[i].used==0)
                          continue;

                      if(bad_tvalues[i].day != iday ||
                          bad_tvalues[i].quart != iquart)
                          continue;

                      bad_tvalues[h].used=0;

		    }

                     /* swap in level 1 data */

                     return(1);
          
		     }

                     else {

                         tdata[iday].stn[j].tlevel2[iquart].qual=1;
                         tdata[iday].stn[j].tlevel2[iquart].data=bad_tvalues[i].fvalue;

		       }

		   }

		}

    }

return(0); 
    
}


void write_bad_tvalues(char *fname,int iday)

{

extern struct bad_daily_values bad_tvalues[6000];
FILE *fp;
char bfile[100],ibuf[100];
char message[GAGEQC_MESSAGE_LEN];
int i,ier;

strcpy(bfile,fname);

fp=fopen(bfile,"w");

if(fp != NULL)
{
	memset(message, '\0', GAGEQC_MESSAGE_LEN);
	sprintf(message,"Opened file: %s\n",bfile);
	logMessage(message);

for(i=0;i<6000;i++) {
         
        if(bad_tvalues[i].used==0)
                        continue;

        if(iday==bad_tvalues[i].day) {

        ier=sprintf(ibuf,"%s %s %d %f\n",bad_tvalues[i].hb5,bad_tvalues[i].parm,
                    bad_tvalues[i].quart,bad_tvalues[i].fvalue);

        fputs(ibuf,fp);

      }

      }

fclose(fp);
}
else
{
	memset(message, '\0', GAGEQC_MESSAGE_LEN);
	sprintf(message,"Could not open file: %s\n",bfile);
	logMessage(message);
}
return;

}

 

void update_bad_tvalues(int iday)

{

extern struct station * tstation;
extern int max_tstations;
extern struct bad_daily_values bad_tvalues[6000];
extern struct tdata tdata[10];
int i,j,h,k;
char parm[10];

for(i=0;i<6000;i++) {

      if(bad_tvalues[i].used==0)
                          continue;

      if(bad_tvalues[i].day != iday)
                          continue;

      bad_tvalues[i].used=0;

   }

for(j=0;j<max_tstations;j++) {

        for(k=0;k<6;k++) {

             if(tdata[iday].stn[j].tlevel2[k].qual != 1)
                                          continue;

             for(h=0;h<6000;h++) {

                      if(bad_tvalues[h].used!=0)
                                     continue;

                      bad_tvalues[h].used=1;
                                          
                      bad_tvalues[h].fvalue=tdata[iday].stn[j].tlevel1[k].data;
                      
                      strcpy(bad_tvalues[h].hb5,tstation[j].hb5);
		      
		      strcpy(parm,tstation[j].parm);
		      
		      if(k<4) 		      
		         parm[5]='Z';
			 
		      else if(k==4) 		      
		         parm[5]='X';
		        
		      if(k==6) 		      
		         parm[5]='N'; 
			 
                      strcpy(bad_tvalues[h].parm,parm);
                      bad_tvalues[h].day=iday;
                      bad_tvalues[h].quart=k;
     
                      break;

		    }
	   }

      }

return;

}

        

void restore_bad_tvalues(int iday,
		         struct station * tstation,
	                 int max_tstations)

{

extern struct tdata tdata[10];
extern struct bad_daily_values bad_tvalues[6000];
int i,j,k;

for(k=0;k<6;k++) {

for(i=0;i<6000;i++) {

      if(bad_tvalues[i].used==0)
                          continue;

      if(bad_tvalues[i].day != iday ||
         bad_tvalues[i].quart != k)
                          continue;

        for(j=0;j<max_tstations;j++) {

            if(strcmp(bad_tvalues[i].hb5,tstation[j].hb5)==0 && 
               bad_tvalues[i].parm[4]==tstation[j].parm[4]) {

                      tdata[iday].stn[j].tlevel2[k].data=bad_tvalues[i].fvalue;
                      tdata[iday].stn[j].tlevel2[k].qual=1;

                      break;

                           }

		    }
      }


}

return;

}
         

int is_tbad(int iday,int iquart,char *hb5,char *parm)

{

extern int max_tstations;
extern struct bad_daily_values bad_tvalues[6000];
int i,j;

for(i=0;i<6000;i++) {

      if(bad_tvalues[i].used==0)
                          continue;

      if(bad_tvalues[i].day != iday ||
         bad_tvalues[i].quart != iquart)
                          continue;
	
        for(j=0;j<max_tstations;j++) {

            if(strcmp(bad_tvalues[i].hb5,hb5)==0 && 
               bad_tvalues[i].parm[4]==parm[4]) 
                       return(1);
 
 		    }
      }

return(0);

}

        

void post_bad_tvalues(int iday)

{
extern struct station * tstation;
extern struct tdata tdata[10];
extern int max_tstations;
extern struct bad_daily_values bad_daily_values[6000];
extern struct bad_daily_values bad_tvalues[6000];

int i,j,k;

for(k=0;k<6;k++) {

for(i=0;i<6000;i++) {

      if(bad_tvalues[i].used==0)
                          continue;

      if(bad_tvalues[i].day != iday ||
         bad_tvalues[i].quart != k)
                          continue;

        for(j=0;j<max_tstations;j++) {

            if(strcmp(bad_tvalues[i].hb5,tstation[j].hb5)==0 && 
               bad_tvalues[i].parm[4]==tstation[j].parm[4]) {

                  if(tdata[iday].stn[j].tlevel2[k].data==bad_tvalues[i].fvalue){

                          tdata[iday].stn[j].tlevel2[k].data=
                                    bad_tvalues[i].fvalue;
        
                          tdata[iday].stn[j].tlevel2[k].qual=1;
        
			}

                      break;

		}

	  }
    }

}
return;


/*  ==============  Statements containing RCS keywords:  */
{static char rcs_id1[] = "$Source$";
 static char rcs_id2[] = "$Id$";}
/*  ===================================================  */

}
  
              

File: cave/com.raytheon.viz.mpe/src/com/raytheon/viz/mpe/util/BadTValues.java
/**
 * This software was developed and / or modified by Raytheon Company,
 * pursuant to Contract DG133W-05-CQ-1067 with the US Government.
 * 
 * U.S. EXPORT CONTROLLED TECHNICAL DATA
 * This software product contains export-restricted data whose
 * export/transfer/disclosure is restricted by U.S. law. Dissemination
 * to non-U.S. persons whether in the United States or abroad requires
 * an export license or other authorization.
 * 
 * Contractor Name:        Raytheon Company
 * Contractor Address:     6825 Pine Street, Suite 340
 *                         Mail Stop B8
 *                         Omaha, NE 68106
 *                         402.291.0100
 * 
 * See the AWIPS II Master Rights File ("Master Rights File.pdf") for
 * further licensing information.
 **/
package com.raytheon.viz.mpe.util;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

import com.raytheon.uf.common.status.IUFStatusHandler;
import com.raytheon.uf.common.status.UFStatus;
import com.raytheon.uf.common.status.UFStatus.Priority;
import com.raytheon.viz.mpe.util.DailyQcUtils.Bad_Daily_Values;
import com.raytheon.viz.mpe.util.DailyQcUtils.Station;
import com.raytheon.viz.mpe.util.DailyQcUtils.Tdata;

/**
 * Bad Temperature Value class.
 * 
 * <pre>
 * 
 * SOFTWARE HISTORY
 * Date         Ticket#    Engineer    Description
 * ------------ ---------- ----------- --------------------------
 * Mar 9, 2009            snaples     Initial creation
 * May 21, 2018   7131    mduff       Updated to accommodate changes to other classes and cleanup.
 * Jul 12, 2018  7357     smanoj      Fixed bad file permission issue.
 * 
 * </pre>
 * 
 * @author snaples
 */

public class BadTValues {
    private final IUFStatusHandler statusHandler = UFStatus
            .getHandler(BadTValues.class);


    private DailyQcUtils dqc = DailyQcUtils.getInstance();

    public void read_bad_tvalues(String tpointd, int m) {

        Bad_Daily_Values bad_tvalues[] = dqc.bad_tvalues;
        Scanner s = null;

        File badTFile = new File(tpointd);
        if (badTFile.exists()) {
            try (BufferedReader in = new BufferedReader(new FileReader(tpointd))) {
                for (int i = 0; i < 6000; i++) {

                    if (bad_tvalues[i].used == 1) {
                        continue;
                    }

                    String vals = in.readLine();

                    if (vals == null) {
                        break;
                    }

                    s = new Scanner(vals);

                    bad_tvalues[i].used = 1;
                    bad_tvalues[i].hb5 = s.next();
                    bad_tvalues[i].parm = s.next();
                    bad_tvalues[i].day = m;
                    bad_tvalues[i].quart = (int) s.nextDouble();
                    bad_tvalues[i].fvalue = s.nextFloat();
                }
                in.close();
            } catch (IOException e) {
            	statusHandler.handle(Priority.ERROR, e.getMessage(), e);
                return;
            }
        }
    }

    public void purge_bad_tvalues(int iday) {
        Bad_Daily_Values bad_tvalues[] = dqc.bad_tvalues;

        for (int k = 0; k < 6; k++) {
            for (int i = 0; i < 6000; i++) {
                if (bad_tvalues[i].used == 1 && bad_tvalues[i].day == iday
                        && bad_tvalues[i].quart == k) {
                    bad_tvalues[i].used = 0;
                }
            }
        }
        return;
    }

    public int get_bad_tvalues(int iday, int iquart) {
        Bad_Daily_Values bad_tvalues[] = dqc.bad_tvalues;
        ReadTemperatureStationList rt = new ReadTemperatureStationList();

        for (int i = 0; i < 6000; i++) {
            if (bad_tvalues[i].used == 0) {
                continue;
            }

            if (bad_tvalues[i].day != iday || bad_tvalues[i].quart != iquart) {
                continue;
            }

            for (int j = 0; j < rt.getNumTstations(); j++) {
                if (bad_tvalues[i].hb5
                        .equals(dqc.temperature_stations.get(j).hb5)
                        && bad_tvalues[i].parm
                                .charAt(4) == dqc.temperature_stations
                                        .get(j).parm.charAt(4)) {

                    if (dqc.tdata[iday].tstn[j].tlevel2[iquart].qual == 5
                            && bad_tvalues[i].fvalue != dqc.tdata[iday].tstn[j].tlevel1[iquart].data
                            && dqc.tdata[iday].tstn[j].tlevel1[iquart].data >= 0) {

                        /* eliminate all bad values for current month */
                        for (int h = 0; h < 6000; h++) {

                            if (bad_tvalues[i].used == 0) {
                                continue;
                            }

                            if (bad_tvalues[i].day != iday
                                    || bad_tvalues[i].quart != iquart) {
                                continue;
                            }

                            bad_tvalues[h].used = 0;
                        }

                        /* swap in level 1 data */

                        return 1;

                    }

                    else {

                        dqc.tdata[iday].tstn[j].tlevel2[iquart].qual = 1;
                        dqc.tdata[iday].tstn[j].tlevel2[iquart].data = bad_tvalues[i].fvalue;
                    }
                }
            }
        }
        return 0;
    }

    public void write_bad_tvalues(String fname, int iday) {
        String ibuf;
        File bfile = new File(fname);
        Bad_Daily_Values bad_tvalues[] = dqc.bad_tvalues;

        try (BufferedWriter out = new BufferedWriter(new FileWriter(bfile))) {

            for (int i = 0; i < 6000; i++) {

                if (bad_tvalues[i].used == 0) {
                    continue;
                }

                if (iday == bad_tvalues[i].day) {
                    ibuf = String.format("%s %s %d %f", bad_tvalues[i].hb5,
                            bad_tvalues[i].parm, bad_tvalues[i].quart,
                            bad_tvalues[i].fvalue);
                    out.write(ibuf);
                    out.newLine();
                }
            }
        }catch (IOException e) {
            statusHandler.handle(Priority.ERROR,
                    "Problem writing bad file " + fname + e.getMessage(), e);
        } finally {

            try{
                // Open up the write permission of the bad file to the GROUP.
                File badfile = new File(fname);
                if (!(Files.getPosixFilePermissions(badfile.toPath()))
                        .equals(FilePermissionHelper.POSIX_FILE_SET)) {
                    FilePermissionHelper.applyFilePermissions(badfile.toPath(),
                            FilePermissionHelper.POSIX_FILE_SET);
                }
            } catch (IOException e) {
                statusHandler.handle(Priority.ERROR,
                        "Failed to update permissions on bad file " + fname
                                + e.getMessage(),
                        e);
           }
        }

        return;
    }

    public void update_bad_tvalues(int iday) {
        char parm[] = new char[10];
        Bad_Daily_Values bad_tvalues[] = dqc.bad_tvalues;
        ReadTemperatureStationList rt = new ReadTemperatureStationList();

        for (int i = 0; i < 6000; i++) {
            if (bad_tvalues[i].used == 0) {
                continue;
            }

            if (bad_tvalues[i].day != iday) {
                continue;
            }

            bad_tvalues[i].used = 0;
        }

        for (int j = 0; j < rt.getNumTstations(); j++) {
            for (int k = 0; k < 6; k++) {
                if (dqc.tdata[iday].tstn[j].tlevel2[k].qual != 1) {
                    continue;
                }

                for (int h = 0; h < 6000; h++) {
                    if (bad_tvalues[h].used != 0) {
                        continue;
                    }

                    bad_tvalues[h].used = 1;

                    bad_tvalues[h].fvalue = dqc.tdata[iday].tstn[j].tlevel1[k].data;

                    bad_tvalues[h].hb5 = dqc.temperature_stations.get(j).hb5;

                    parm = dqc.temperature_stations.get(j).parm.toCharArray();

                    if (k < 4) {
                        parm[5] = 'Z';
                    } else if (k == 4) {
                        parm[5] = 'X';
                    }

                    if (k == 6) {
                        parm[5] = 'N';
                    }

                    bad_tvalues[h].parm = Arrays.toString(parm);
                    bad_tvalues[h].day = iday;
                    bad_tvalues[h].quart = k;

                    break;
                }
            }
        }
        return;
    }

    public void restore_bad_tvalues(int iday,
            List<Station> temperature_stations, int max_tstations) {
        for (int k = 0; k < 6; k++) {
            for (int i = 0; i < 6000; i++) {
                if (dqc.bad_tvalues[i].used == 0) {
                    continue;
                }

                if (dqc.bad_tvalues[i].day != iday
                        || dqc.bad_tvalues[i].quart != k) {
                    continue;
                }

                for (int j = 0; j < max_tstations; j++) {
                    if ((dqc.bad_tvalues[i].hb5
                            .equalsIgnoreCase(temperature_stations.get(j).hb5)
                            && dqc.bad_tvalues[i].parm.charAt(
                                    4) == temperature_stations.get(j).parm
                                            .charAt(4))) {

                        dqc.tdata[iday].tstn[j].tlevel2[k].data = dqc.bad_tvalues[i].fvalue;
                        dqc.tdata[iday].tstn[j].tlevel2[k].qual = 1;

                        break;
                    }
                }
            }
        }

        return;
    }

    public int is_tbad(int iday, int iquart, String hb5, String parm) {
        Bad_Daily_Values bad_tvalues[] = dqc.bad_tvalues;
        ReadTemperatureStationList rt = new ReadTemperatureStationList();

        for (int i = 0; i < 6000; i++) {
            if (bad_tvalues[i].used == 0) {
                continue;
            }

            if (bad_tvalues[i].day != iday || bad_tvalues[i].quart != iquart) {
                continue;
            }

            for (int j = 0; j < rt.getNumTstations(); j++) {
                if (bad_tvalues[i].hb5.equals(hb5)
                        && bad_tvalues[i].parm.charAt(4) == parm.charAt(4)) {
                    return 1;
                }
            }
        }

        return 0;
    }

    public void post_bad_tvalues(int iday) {
        Bad_Daily_Values bad_tvalues[] = dqc.bad_tvalues;
        Tdata tdata[] = dqc.tdata;
        ReadTemperatureStationList rt = new ReadTemperatureStationList();

        for (int k = 0; k < 6; k++) {
            for (int i = 0; i < 6000; i++) {
                if (bad_tvalues[i].used == 0) {
                    continue;
                }

                if (bad_tvalues[i].day != iday || bad_tvalues[i].quart != k) {
                    continue;
                }

                for (int j = 0; j < rt.getNumTstations(); j++) {
                    if ((bad_tvalues[i].hb5.equals(
                            dqc.temperature_stations.get(j).hb5))
                            && bad_tvalues[i].parm.charAt(
                                    4) == dqc.temperature_stations.get(j).parm
                                            .charAt(4)) {

                        if (tdata[iday].tstn[j].tlevel2[k].data == bad_tvalues[i].fvalue) {
                            tdata[iday].tstn[j].tlevel2[k].data = bad_tvalues[i].fvalue;
                            tdata[iday].tstn[j].tlevel2[k].qual = 1;
                        }

                        break;
                    }
                }
            }
        }
        return;
    }
}
